File: ruby/153.find-minimum-in-rotated-sorted-array.rb
# @param {Integer[]} nums
# @return {Integer}
def find_min(nums)
  left = 0
  right = nums.length - 1

  return nums[left] if nums[left] < nums[right]

  while left + 1 < right
    mid = (left + right) / 2

    if nums[mid] > nums[left]
      left = mid
    else
      right = mid
    end
  end

  [nums[left], nums[right]].min
end

File: rust/src/153.find-minimum-in-rotated-sorted-array.rs
impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let mut left = 0;
        let mut right = nums.len() - 1;

        if nums[left] < nums[right] {
            return nums[left];
        }

        while left < right {
            let mid = (left + right) / 2;

            if nums[0] > nums[mid] {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        nums[left]
    }
}
